File: ts/256/235.ts
import { TreeNode } from '../class'

function lowestCommonAncestor(root: TreeNode, p: TreeNode, q: TreeNode): TreeNode | null {
  if (p.val > q.val) return lowestCommonAncestor(root, q, p)
  let node: TreeNode | null = root
  while (node) {
    if (node.val >= p.val && node.val <= q.val) return node
    if (node.val <= p.val) {
      node = node.right
    } else if (node.val >= q.val) {
      node = node.left
    }
  }
  return node
};

File: py3/256/235.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if p.val > q.val: return self.lowestCommonAncestor(root, q, p)
        ptr = root
        while ptr:
            if ptr.val < p.val: ptr = ptr.right
            elif ptr.val > q.val: ptr = ptr.left
            else: break
        return ptr